File: 5.longest-palindromic-substring.js
// @leet start
/**
 * @param {string} s
 * @return {string}
 */
var checkpal = function(l, ans, r, s) {
	while (l >= 0 && r < s.length) {
		if (s[l] === s[r]) {
			ans = s[l] + ans + s[r];
			l--;
			r++;
		} else {
			return ans;
		}
	}
	return ans;
};
var longestPalindrome = function(s) {
	let realans = "";
	let ans = "";
	for (let i = 0; i < s.length; i++) {
		ans = s[i];
		let l = i - 1;
		let r = i + 1;
		let check = checkpal(l, ans, r, s);
		if (check.length > realans.length) {
			realans = check;
		}
		ans = "";
		l = i;
		r = i + 1;
		check = checkpal(l, ans, r, s);
		if (check.length > realans.length) {
			realans = check;
		}
	}
	return realans;
};
// @leet end

File: 5.longest-palindromic-substring.cpp
// @leet start
#include <string>
using namespace std;
class Solution {
public:
  bool ispal(string s, int start, int end) {
    while (start < end) {
      if (s[start] != s[end]) {
        return false;
      }
      start++;
      end--;
    };
    return true;
  }
  string longestPalindrome(string s) {
    if (s.size() <= 1) return s;
    string answer = string(1,s[0]);
    int maxlength = 0;
    for (int i = 0; i < s.size(); i++) {
      for (int j = i + 1; j < s.size(); j++) {
        if (s[j] == s[i] && j - i +1> maxlength) {
          if (ispal(s, i, j)) {
            answer = s.substr(i, j-i+1);
						maxlength = answer.size();
          }
        }
      }
    }
    return answer;
  }
};
// @leet end
